README
No more forgotten football matches or formula one races! Avoid discussion with
your better half and let kodi switch to the channels you want! Or let they switch
to their favorite shows at a given time!

This service adds items to the context menu of the pvr osd guide window (PVROSDGuide)
automatically (add switchtimer, delete all switchtimer). Currently the skins Aeon Flex,
Destiny and a Confluence Mod from Kodinerds fully supports this service with additional
windows but other skins can this service implement too. Hints for integration see 
below here. If you aren't a skinner or don't need special windows for editing/deleting 
switchtimers just ignore the skinners section and install this service as usual. All can 
be done with the context menu of pvr osd guide.

Skinners only:

If you want or need to call this service inside of another window different from 
PVROSDGuide, use:

<label>$ADDON[service.kn.switchtimer 30040]</label>
<visible>System.HasAddon(service.kn.switchtimer) + Window.IsVisible(tvguide)</visible>
<onclick>RunScript(service.kn.switchtimer,action=add,channel=channelname,icon=icon,date=datestring,title=title)</onclick>

You have to provide the needed parameters, comparsion is done via valid Channelname and 
Datetime. Format of Datetime must be the same as used in Kodi settings:

channelname (e.g. $INFO[ListItem.Channelname])
icon (e.g. $INFO[ListItem.Icon])
datestring (e.g. $INFO[ListItem.Date])
title (e.g. $INFO[ListItem.Title])

If you want buttons for deleting one or all timers just use:

<onclick>RunScript(service.kn.switchtimer,action=del,timer=tx)</onclick>

where tx is a timer from t0 to t9, or delete all timers with

<label>$ADDON[service.kn.switchtimer 30041]</label>
<onclick>RunScript(service.kn.switchtimer,action=delall)</onclick>

Timers are stored as strings in the settings of the skin (guisettings.xml/skinsettings.xml)
as follows. Empty strings are inactive timers. All timers are sorted by datetime
where t0 is the nearest and t9 the latest timer.

t0:channel      # channel name
t0:icon         # channel icon
t0:date         # datetime of broadcast
t0:title        # title of broadcast
...
t9:channel
t9:icon
t9:date
t9:title

SwitchTimerActiveItems: # of active timers


changelog.txt
CHANGELOG

2.0.10
- special notification window added

2.0.9
- Timerlist for showing and deleting timers, adding leading time offset in settings

2.0.8
- Option for channel switching at startup added

2.0.6
- Cosmetics

2.0.5
- Bugfixes

2.0.4
- Portugese (brazilian) language added

2.0.3
- Module dateutil included.

2.0.2
- create settings path if not exists. This avoids IOExceptions when writing timers and the addon
  wasn't configured yet.

2.0.1
- Profile paths for timer.json considered

2.0.0
- complete Redesign of source code
- Timers are stored into a JSON structured file independent from skin
- IMPORTANT: Timers use now skin properties instead of skin labels

1.2.3
- Dutch translation

1.2.2
- Bug in countdown dialog fixed

1.2.1
- merge from pull request of W. Heisenberg:
  - Countdown dialog added
  - Notification or countdown dialog selectable

1.2.0
- adding context menu items of PVROSDGuide using the extension
  point kodi.context.item of Isengard

1.1.2
- Internationalization even for skin labels

1.1.1
- Channel switch could be aborted by pressing any key while
  switch message is displayed on screen

1.1.0
- complete rewrite, refactoring and optimization of code
- timer t0-t9 now sorted by datetime (t0: nearest)

1.0.12
- Channelicon property added (t0:icon ... t9:icon)

1.0.11
- some JSONRPC Errors fixed
- changelog added

1.0.0
- intial commit


handler.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import time
import xbmc, xbmcaddon, xbmcgui
import os
import operator
import json

import resources.lib.knClasses as knClasses

addon = xbmcaddon.Addon()
addonid = xbmcaddon.Addon().getAddonInfo('id')
addonname = xbmcaddon.Addon().getAddonInfo('name')
path = xbmc.translatePath(xbmcaddon.Addon().getAddonInfo('path'))
profiles = xbmc.translatePath(xbmcaddon.Addon().getAddonInfo('profile'))
version = xbmcaddon.Addon().getAddonInfo('version')
loc = xbmcaddon.Addon().getLocalizedString

IconDefault = os.path.join(path, 'resources', 'media', 'default.png')
IconAlert = os.path.join(path, 'resources', 'media', 'alert.png')
IconOk = os.path.join(path, 'resources', 'media', 'ok.png')

confirmTmrAdded = True if xbmcaddon.Addon().getSetting('confirmTmrAdded').upper() == 'TRUE' else False

OSD = xbmcgui.Dialog()
OSDProgress = xbmcgui.DialogProgress()
HOME = xbmcgui.Window(10000)

if not os.path.exists(profiles): os.makedirs(profiles, 0755)
__timer__ = os.path.join(profiles, 'timer.json')

__timerdict__ = {'channel': None, 'icon': None, 'date': None, 'title': None, 'plot': None}

def putTimer(timers):
    for timer in timers:
        if not timer['utime'] or timer['utime'] < time.time(): timers.remove(timer)
    with open(__timer__, 'w') as handle:
        json.dump(timers, handle)
    HOME.setProperty('SwitchTimerActiveItems', str(len(timers)))
    notifyLog('%s timer(s) written' % (len(timers)), xbmc.LOGNOTICE)

def getTimer():
    try:
        with open(__timer__, 'r') as handle:
            timers = json.load(handle)
    except IOError:
        timers = []
    HOME.setProperty('SwitchTimerActiveItems', str(len(timers)))
    return timers

def getSetting(setting):
    return xbmcaddon.Addon().getSetting(setting)

def notifyLog(message, level=xbmc.LOGDEBUG):
    xbmc.log('[%s] %s' % (addonid, message.encode('utf-8')), level)

def notifyOSD(header, message, icon=IconDefault, time=5000):
    OSD.notification(header.encode('utf-8'), message.encode('utf-8'), icon, time)

def setTimer(params):
    utime = knClasses.date2timeStamp(params['date'])
    if not utime: return False

    _now = int(time.time())
    if _now > utime:
        notifyLog('Timer date in the past', xbmc.LOGNOTICE)
        notifyOSD(loc(30000), loc(30022), icon=IconAlert)
        return False

    timers = getTimer()
    for timer in timers:
        if knClasses.date2timeStamp(timer['date']) == utime:
            notifyLog('Timer already set, ask for replace', xbmc.LOGNOTICE)
            _res = OSD.yesno(addonname, loc(30031) % (timer['channel'], timer['title']))

            if not _res: return False
            timers.remove(timer)

    if len(timers) > 9:
        notifyLog('Timer limit exceeded, no free slot', xbmc.LOGFATAL)
        notifyOSD(loc(30000), loc(30024), icon=IconAlert)
        return False

    # append timer and sort timerlist

    params['utime'] = utime
    timers.append(params)
    timers.sort(key=operator.itemgetter('utime'))

    setTimerProperties(timers)
    putTimer(timers)

    notifyLog('Timer added @%s, %s, %s' % (params['date'], params['channel'].decode('utf-8'), params['title'].decode('utf-8')), xbmc.LOGNOTICE)
    if confirmTmrAdded: notifyOSD(loc(30000), loc(30021), icon=IconOk)
    return True

def setTimerProperties(timerlist):
    _idx = 0
    for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']:

        if _idx < len(timerlist):
            # Set timer properties
            for element in __timerdict__:
                try:
                    HOME.setProperty('%s:%s' % (prefix, element), timerlist[_idx][element])
                except KeyError:
                    pass
            _idx += 1
        else:
            # Clear remaining properties
            clearTimerProperties(prefix)

def clearTimerProperties(prefix=None):
    if not prefix:
        for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']:
            for element in __timerdict__: HOME.clearProperty('%s:%s' % (prefix, element))
        timers = []
        notifyLog('Properties of all timers deleted, timerlist cleared')
    else:
        _date = HOME.getProperty('%s:date' % (prefix))
        if _date == '': return False
        timers = getTimer()
        for timer in timers:
            if timer['date'] == _date: timer['utime'] = None
            for element in __timerdict__: HOME.clearProperty('%s:%s' % (prefix, element))
            notifyLog('Properties for timer %s @%s deleted' % (prefix, _date))

    putTimer(timers)
    return True

if __name__ ==  '__main__':

    notifyLog('Parameter handler called')
    try:
        if sys.argv[1]:
            args = {'action':None, 'channel':'', 'icon': '', 'date':'', 'title':'', 'plot': ''}
            pars = sys.argv[1:]
            for par in pars:
                try:
                    item, value = par.split('=')
                    args[item] = value.replace(',', '&comma;').decode('utf-8')
                    notifyLog('Provided parameter %s: %s' % (item, args[item]))
                except ValueError:
                    args[item] += ', ' + par

            if args['action'] == 'add':
                if not setTimer(args):
                    notifyLog('Timer couldn\'t or wouldn\'t set', xbmc.LOGERROR)
            elif args['action'] == 'del':
                clearTimerProperties(args['timer'])
            elif args['action'] == 'delall':
                for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']: clearTimerProperties()
    except IndexError:
        _tlist = []
        timers = getTimer()
        if len(timers) > 0:
            for timer in timers:
                liz = xbmcgui.ListItem(label='%s: %s' % (timer['date'], timer['channel']),
                                       label2=timer['title'], iconImage=timer['icon'])
                liz.setProperty('utime', str(timer['utime']))
                liz.setProperty('date', timer['date'])
                _tlist.append(liz)

            _idx = OSD.select(loc(30015), _tlist, useDetails=True)
            if _idx > -1:
                timers[_idx].update({'utime': None})
                _date = timers[_idx].get('date', '')
                for element in __timerdict__: HOME.clearProperty('t%s:%s' % (_idx, element))
                putTimer(timers)
                notifyLog('Properties for timer t%s @%s deleted' % (_idx, _date))

        else:
            notifyLog('No active timers yet', xbmc.LOGERROR)
            OSD.ok(loc(30000), loc(30030))
    except Exception, e:
        notifyLog('Script error: %s' % (e.message), xbmc.LOGERROR)



resources/lib/knClasses.py
import xbmc
import xbmcgui
import json
import time, datetime

EPOCH = datetime.datetime(1970, 1, 1)
JSON_TIME_FORMAT = '%Y-%m-%d %H:%M:%S'
UTC_OFFSET = int(round((datetime.datetime.now() - datetime.datetime.utcnow()).seconds, -1))

def date2timeStamp(date, dFormat=None, utc=False):
    # Kodi bug: returns '%H%H' or '%I%I'

    if dFormat is None:
        # use Kodi's own dateformat provided by setup
        df = xbmc.getRegion('dateshort') + ' ' + xbmc.getRegion('time').replace('%H%H', '%H').replace('%I%I','%I').replace(':%S', '')
    else:
        df = dFormat
    dtt = time.strptime(date, df)
    if not utc: return int(time.mktime(dtt))
    return int(time.mktime(dtt)) + UTC_OFFSET


class XBMCMonitor(xbmc.Monitor):

    def __init__(self, *args, **kwargs):
        xbmc.Monitor.__init__(self)
        self.SettingsChanged = False

    def onSettingsChanged(self):
        self.SettingsChanged = True


class cNotification(xbmcgui.WindowXMLDialog):

    TEXT = 1151
    PROGRESS = 1155
    THUMB = 1156
    BUTTON_RECORD = 1152
    BUTTON_SWITCH = 1153
    BUTTON_CANCEL = 1154

    ACTION_SELECT = 7
    ACTION_NAV_BACK = 92


    def __init__(self, *args, **kwargs):

        self.timer = int(kwargs.get('timer', 5))
        self.text = kwargs.get('message', '')
        self.icon = kwargs.get('icon', '')
        self.recEnabled = kwargs.get('recEnabled', True)

        self.isCanceled = False
        self.requestSwitch = False
        self.initRecording = False

    def onInit(self):
        self.textControl = self.getControl(self.TEXT)  # Notification text field
        self.progressControl = self.getControl(self.PROGRESS)  # Notification progress bar
        self.imageControl = self.getControl(self.THUMB)  # Notification channel thumb
        self.buttonControl_1 = self.getControl(self.BUTTON_RECORD)  # Button 1 (Record)
        self.buttonControl_2 = self.getControl(self.BUTTON_SWITCH)  # Button 2 (Switch)
        self.buttonControl_3 = self.getControl(self.BUTTON_CANCEL)  # Button 3 (Cancel)

        self.textControl.setText(self.text)
        self.imageControl.setImage(self.icon)
        self.buttonControl_1.setEnabled(self.recEnabled)

        for t in range(100, -1, -1):
            self.progressControl.setPercent(t)
            xbmc.sleep(self.timer * 10)
            if self.isCanceled or self.requestSwitch or self.initRecording: break
        self.close()

    def onClick(self, controlId):
        if controlId == self.BUTTON_RECORD: self.initRecording = True
        elif controlId == self.BUTTON_SWITCH: self.requestSwitch = True
        elif controlId == self.BUTTON_CANCEL: self.isCanceled = True
        self.close()

    def onAction(self, action):
        if action == self.ACTION_NAV_BACK: self.close()


    def set_text(self, p):
        self.textControl.setText(p)


    def close(self):
        xbmcgui.WindowXMLDialog.close(self)


class cPvrProperties(object):

    class PvrAddTimerException(Exception):
        pass

    class JsonExecException(Exception):
        pass


    def __init__(self, *args, **kwargs):
        pass

    def jsonrpc(self, query):
        querystring = {"jsonrpc": "2.0", "id": 1}
        querystring.update(query)
        try:
            response = json.loads(xbmc.executeJSONRPC(json.dumps(querystring, encoding='utf-8')))
            if 'result' in response: return response['result']
        except TypeError, e:
            raise self.JsonExecException('Error executing JSON RPC: %s' % e.message)
        return None

    def getRecordingCapabilities(self, pvrid, timestamp):
        '''
        :param pvrid:       str PVR-ID of the broadcast station
        :param timestamp:   int UNIX timestamp
        :return:            dict: int unique broadcastID of the broadcast or None, bool hastimer
        '''
        params = {'broadcastid': None, 'hastimer': False}
        query = {
            "method": "PVR.GetBroadcasts",
            "params": {"channelid": pvrid,
                       "properties": ["title", "starttime", "hastimer"]}
        }
        res = self.jsonrpc(query)
        try:
            for broadcast in res.get('broadcasts'):
                if timestamp == date2timeStamp(broadcast['starttime'], dFormat=JSON_TIME_FORMAT, utc=True):
                    params.update({'broadcastid': broadcast['broadcastid'], 'hastimer': broadcast['hastimer']})
                    return params
            raise self.PvrAddTimerException('No broadcast found for pvr ID %s@%s' % (pvrid, timestamp))
        except (TypeError, AttributeError), e:
            raise self.PvrAddTimerException('Error on determining broadcast for pvr ID %s: %s' % (pvrid, e.message))

    def setTimer(self, broadcastId):
        '''
        :param broadcastId: int unique broadcastID of the broadcast
        :return:            none
        '''
        query = {
            "method": "PVR.AddTimer",
            "params": {"broadcastid": broadcastId}
        }
        res = self.jsonrpc(query)
        if res == 'OK': return
        else:
            raise self.PvrAddTimerException('Timer for braodcast %s couldn\'t added' % broadcastId)

service.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import time
import xbmc, xbmcaddon, xbmcgui
import json
import os
import re

import handler
import resources.lib.knClasses as knClasses

addon = xbmcaddon.Addon()
addonid = addon.getAddonInfo('id')
addonname = addon.getAddonInfo('name')
path = xbmc.translatePath(addon.getAddonInfo('path'))
version = addon.getAddonInfo('version')
loc = addon.getLocalizedString

IconDefault = os.path.join(path, 'resources', 'media', 'default.png')
IconAlert = os.path.join(path, 'resources', 'media', 'alert.png')
IconOk = os.path.join(path, 'resources', 'media', 'ok.png')

INTERVAL = 10 # More than that will make switching too fuzzy because service isn't synchronized with real time
HOME = xbmcgui.Window(10000)

SKIN = xbmc.translatePath('special://skin').split(os.sep)[-2] + '.st_notification.xml'

def jsonrpc(query):
    querystring = {"jsonrpc": "2.0", "id": 1}
    querystring.update(query)
    return json.loads(xbmc.executeJSONRPC(json.dumps(querystring, encoding='utf-8')))


class Service(knClasses.XBMCMonitor):

    def __init__(self, *args):

        knClasses.XBMCMonitor.__init__(self)
        self.skinPath = None

        self.getSettings()
        handler.notifyLog('Init Service %s %s' % (addonname, version))
        self.bootstrap = True
        self.timers = handler.getTimer()
        handler.setTimerProperties(self.timers)

    def getSettings(self):

        # There seems to be a bug in kodi as sometimes changed properties wasn't read/update properly even if
        # monitor signaled a change.
        # Reading of settings is now outsourced to handler as a workaround.

        self.__showNoticeBeforeSw = True if handler.getSetting('showNoticeBeforeSw').upper() == 'TRUE' else False
        self.__useCountdownTimer = True if handler.getSetting('useCountdownTimer').upper() == 'TRUE' else False
        self.__leadOffset = int(re.match('\d+', handler.getSetting('leadOffset')).group())
        self.__dispMsgTime = int(re.match('\d+', handler.getSetting('dispTime')).group())*1000
        self.__discardTmr = int(re.match('\d+', handler.getSetting('discardOldTmr')).group())*60
        self.__confirmTmrAdded = True if handler.getSetting('confirmTmrAdded').upper() == 'TRUE' else False

        self.__switchOnInit = True if handler.getSetting('switchOnInit').upper() == 'TRUE' else False
        try:
            self.__channel = int(re.match('\d+', handler.getSetting('channel')).group())
        except AttributeError:
            self.__channel = 0

        self.SettingsChanged = False

    @classmethod

    def resetTmr(cls, date):
        for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']:
            if HOME.getProperty('%s:date' % (prefix)) == '': continue
            elif HOME.getProperty('%s:date' % (prefix)) == date: handler.clearTimerProperties(prefix)

    @classmethod

    def channelName2channelId(cls, channelname):
        query = {
                "method": "PVR.GetChannels",
                "params": {"channelgroupid": "alltv"},
                }
        res = jsonrpc(query)
        if 'result' in res and 'channels' in res['result']:
            res = res['result'].get('channels')
            for channels in res:
                if channels['label'] == channelname: return channels['channelid']
        return False

    @classmethod

    def getPlayer(cls):
        props = {'player': None, 'playerid': None, 'media': None, 'id': None}
        query = {
                "method": "Player.GetActivePlayers",
                }
        res = jsonrpc(query)
        if 'result' in res and res['result']:
            res = res['result'][0]
            props['player'] = res['type']
            props['playerid'] = res['playerid']

            query = {
                    "method": "Player.GetItem",
                    "params": {"properties": ["title", "season", "episode", "file"],
                               "playerid": props['playerid']},
                    "id": "VideoGetItem"
                    }
            res = jsonrpc(query)
            if 'result' in res:
                res = res['result'].get('item')
                props['media'] = res['type']
                if 'id' in res: props['id'] = res['id']
        return props

    @classmethod

    def switchToChannelId(cls, playerProperties, channelId, channel):

        if playerProperties['player'] == 'audio' or (playerProperties['player'] == 'video' and playerProperties['media'] != 'channel'):

            # stop all other players except pvr

            handler.notifyLog('player:%s media:%s @id:%s is running' %
                              (playerProperties['player'], playerProperties['media'], playerProperties['playerid']))
            query = {
                "method": "Player.Stop",
                "params": {"playerid": playerProperties['playerid']},
            }
            res = jsonrpc(query)
            if 'result' in res and res['result'] == "OK":
                handler.notifyLog('Player stopped')

        handler.notifyLog('Currently playing channelid %s, switch to id %s' % (playerProperties['id'], channelId))
        query = {
            "method": "Player.Open",
            "params": {"item": {"channelid": channelId}}
        }
        res = jsonrpc(query)
        if 'result' in res and res['result'] == 'OK':
            handler.notifyLog('Switched to channel \'%s\'' % (channel))
        else:
            handler.notifyLog('Couldn\'t switch to channel \'%s\'' % (channel))
            handler.notifyOSD(loc(30000), loc(30025) % (channel), icon=IconAlert)


    def poll(self):

        while not knClasses.XBMCMonitor.abortRequested(self):

            if knClasses.XBMCMonitor.waitForAbort(self, INTERVAL): break
            if self.SettingsChanged:
                self.getSettings()

            _now = time.time()
            _switchInstantly = False
            plrProps = self.getPlayer()

            for _timer in self.timers:

                if not _timer['utime']:
                    handler.notifyLog('Couldn\'t calculate timestamp, delete timer', xbmc.LOGERROR)
                    self.resetTmr(_timer['date'])
                    break

                # delete old/discarded timers
                if _timer['utime'] + self.__discardTmr < _now:
                    self.resetTmr(_timer['date'])
                    continue

                if _timer['utime'] - self.__leadOffset < _now: _switchInstantly = True
                if (_timer['utime'] - _now < INTERVAL + self.__dispMsgTime / 1000 + self.__leadOffset) or _switchInstantly:
                    chanIdTmr = self.channelName2channelId(_timer['channel'])
                    if chanIdTmr:

                        # compare with player properties, switch if necessary

                        if chanIdTmr == plrProps['id']:
                            handler.notifyLog('Channel switching unnecessary')
                            handler.notifyOSD(loc(30000), loc(30027) % (_timer['title'], _timer['channel']), time=self.__dispMsgTime)
                        else:
                            switchAborted = False
                            secs = 0
                            handler.notifyLog('Channel switch to %s required' % (_timer['channel']))

                            if _switchInstantly:
                                handler.notifyLog('immediate channel switching required')
                                handler.notifyOSD(loc(30000), loc(30027) % (_timer['title'], _timer['channel']), time=5000)
                            elif not self.__showNoticeBeforeSw: xbmc.sleep(self.__dispMsgTime)
                            elif self.__useCountdownTimer:
                                if os.path.exists(os.path.join(path, 'resources', 'skins', 'Default', '1080i', SKIN)):

                                    try:
                                        pvr = knClasses.cPvrProperties()
                                        pvrprops = dict()
                                        recEnabled = True
                                        pvrprops.update(pvr.getRecordingCapabilities(
                                            self.channelName2channelId(_timer['channel']), _timer['utime']))
                                        if pvrprops['hastimer']: recEnabled = False
                                    except (pvr.PvrAddTimerException, pvr.JsonExecException), e:
                                        handler.notifyLog(str(e), xbmc.LOGERROR)
                                        recEnabled = False

                                    Popup = knClasses.cNotification(SKIN, path, message=loc(30035) % (_timer['title'], _timer['channel']),
                                                          timer=self.__dispMsgTime/1000, icon=_timer['icon'], recEnabled=recEnabled)
                                    Popup.doModal()
                                    if Popup.isCanceled: switchAborted = True
                                    elif Popup.initRecording:
                                        try:
                                            pvr.setTimer(pvrprops['broadcastid'])
                                            handler.notifyLog('Recording schedule set for %s: %s' % (_timer['channel'], _timer['title']), xbmc.LOGNOTICE)
                                            switchAborted = True
                                        except (pvr.PvrAddTimerException, pvr.JsonExecException), e:
                                            handler.notifyOSD(loc(30000), loc(30036) % (_timer['title'], _timer['channel']), icon=IconAlert)
                                            handler.notifyLog(str(e), xbmc.LOGERROR)
                                    elif Popup.requestSwitch:
                                        handler.notifyLog('immediate channel switch initate by user')
                                    else:
                                        handler.notifyLog('Window countdown completed without user action')
                                else:
                                    handler.OSDProgress.create(loc(30028), loc(30026) %
                                                               (_timer['channel'], _timer['title']),
                                                               loc(30029) % (int(self.__dispMsgTime / 1000 - secs)))
                                    while secs < self.__dispMsgTime /1000:
                                        secs += 1
                                        percent = int((secs * 100000) / self.__dispMsgTime)
                                        handler.OSDProgress.update(percent, loc(30026) %
                                                                   (_timer['channel'], _timer['title']),
                                                                   loc(30029) % (int(self.__dispMsgTime / 1000 - secs)))
                                        xbmc.sleep(1000)
                                        if (handler.OSDProgress.iscanceled()):
                                            switchAborted = True
                                            break
                                    handler.OSDProgress.close()
                            else:
                                idleTime = xbmc.getGlobalIdleTime()
                                handler.notifyOSD(loc(30000), loc(30026) %
                                                  (_timer['channel'], _timer['title']), time=self.__dispMsgTime)
                                while secs < self.__dispMsgTime /1000:
                                    if idleTime > xbmc.getGlobalIdleTime():
                                        switchAborted = True
                                        break

                                    xbmc.sleep(1000)
                                    idleTime += 1
                                    secs += 1
                                if switchAborted: handler.notifyOSD(loc(30000), loc(30032))

                            if switchAborted: handler.notifyLog('Channelswitch cancelled by user')
                            else:
                                self.bootstrap = False
                                self.switchToChannelId(plrProps, chanIdTmr, _timer['channel'])

                    self.resetTmr(_timer['date'])

            if self.bootstrap and self.__switchOnInit and self.__channel > 0:
                handler.notifyLog('Channelswitch on startup enabled')
                query = {
                        "method": "PVR.GetChannels",
                        "params": {"channelgroupid": "alltv", "properties": ["channelnumber"]}
                        }
                res = jsonrpc(query)
                if 'result' in res:
                    for _channel in res['result']['channels']:
                        if _channel['channelnumber'] == self.__channel:
                            handler.notifyLog('Channelswitch on startup is enabled, switch to \'%s\'' % (_channel['label']))
                            handler.notifyOSD(loc(30000), loc(30013) % (_channel['label']))
                            self.switchToChannelId(plrProps, _channel['channelid'], _channel['label'])
                            self.bootstrap = False
                            break

            self.timers = handler.getTimer()

        handler.notifyLog('Service kicks off')

service = Service()
service.poll()
del service
